event-store/src/adapters/amqp/emitter.rs

```diff
@@ -0,0 +1,182 @@
+    exchange: String,
+    handler: H,
+    channel: Channel<TcpStream>,
+) -> impl Future<Item = (), Error = io::Error>
+where
+    E: EventData,
+    H: Fn(&Event<E>) -> (),
+{
+    let c_channel = channel.clone();
+    let queue_name1 = queue_name.clone();
+    info!("Creating queue {}", queue_name);
+    channel
+        .queue_declare(
+            &queue_name,
+            QueueDeclareOptions {
+                durable: true,
+                exclusive: false,
+                auto_delete: false,
+                ..QueueDeclareOptions::default()
+            },
+            FieldTable::new(),
+        ).and_then(move |queue| {
+            info!("Binding queue {} to exchange {}", queue_name, exchange);
+            channel
+                .queue_bind(
+                    &queue_name,
+                    &exchange,
+                    &event_name,
+                    QueueBindOptions::default(),
+                    FieldTable::new(),
+                ).and_then(move |_| {
+                    channel.basic_consume(
+                        &queue,
+                        &queue_name,
+                        BasicConsumeOptions::default(),
+                        FieldTable::new(),
+                    )
+                }).and_then(move |stream| {
+                    info!("Starting to consume from queue {}", queue_name1);
+                    stream.for_each(move |message| {
+                        let data: Event<E> =
+                            serde_json::from_str(str::from_utf8(&message.data).unwrap()).unwrap();
+                        info!("Receiving message with id {}", data.id);
+                        handler(&data);
+                        c_channel.basic_ack(message.delivery_tag, false)
+                    })
+                })
+        }).and_then(|_| ok(()))
+        .map_err(|e| e.into())
+}
+
+impl EmitterAdapter for AMQPEmitterAdapter {
+    fn get_subscriptions(&self) -> Vec<String> {
+        vec![]
+    }
+
+    fn emit<'a, E: Events + Sync>(
+        &self,
+        event: &Event<E>,
+    ) -> Box<Future<Item = (), Error = io::Error> + Send + Sync> {
+        let payload: Vec<u8> = serde_json::to_string(event)
+            .expect("Cant serialise event")
+            .into();
+        let event_type = event.data().event_type();
+        let id = event.id;
+        info!("Emiting event {} with id {}", event_type, id);
+
+        Box::new(
+            self.channel
+                .basic_publish(
+                    &self.exchange,
+                    &event_type,
+                    payload,
+                    BasicPublishOptions::default(),
+                    BasicProperties::default(),
+                ).and_then(move |_| {
+                    info!("Event with id {} delivered", id);
+                    ok(())
+                }),
+        )
+    }
+
+    fn subscribe<ED, H>(&mut self, handler: H) -> Box<Future<Item = (), Error = io::Error> + Send>
+    where
+        ED: EventData + 'static,
```

Comment:
Hmm. Why is this `'static`? Can you add a lifetime to `AMQPEmitterAdapter` and use it here instead of `'static`?

Reply:
Actually is a bit complicated. The handlers run in the [default tokio executor](https://docs.rs/tokio/0.1.8/tokio/executor/fn.spawn.html), because the the `AMQPEmitterAdapter` is lazy, it delegates the choice of the executor to instantiator and we do not know at compile time the lifetimes of it, it will usually run for as long as the process runs.

At least that is my current understanding, but I may be getting it wrong.